lib/decompress/zstd_decompress_block.c

```diff
@@ -1161,20 +1217,73 @@ ZSTD_decompressSequences_body( ZSTD_DCtx* dctx,
         __asm__(".p2align 4");
 #  endif
 #endif
-        for ( ; ; ) {
-            seq_t const sequence = ZSTD_decodeSequence(&seqState, isLongOffset);
-            size_t const oneSeqSize = ZSTD_execSequence(op, oend, sequence, &litPtr, litEnd, prefixStart, vBase, dictEnd);
+        /* decompress without overrunning litPtr begins */
+        if (litInDst)
+        {
+            for ( ; ; ) {
+                seq_t sequence = ZSTD_decodeSequence(&seqState, isLongOffset);
+                size_t oneSeqSize;
+                if (litPtr + sequence.litLength > dctx->litBufferEnd)
```

Comment:
One thing that might be helpful is looking at the perf annotations, with `perf record --call-graph=dwarf -e cycles:ppp -- ./zstd [...]`. I see the following here:
<img width="663" alt="Screen Shot 2021-08-24 at 12 26 50 PM" src="https://user-images.githubusercontent.com/17059792/130654490-7f83e998-8a83-4c10-b5d8-d634092a019e.png">

Which means this branch is maybe the culprit for some extra cycles? If I annotate this condition with `LIKELY` then for some reason I get a 5-6% speedup in decompression. So maybe it's worth investigating this branch/loop in particular?